### ISSUE TITLE
Detailed Specifications of Automatic Updates

### ISSUE THREAD

**Issue by Hiroki-Aoki** [serverless/serverless#12866]
### Issue description

Please tell me the detailed specifications of the automatic updates explained in https://github.com/serverless/serverless?tab=readme-ov-file#update-serverless-framework.

# Questions
1. Is there a way to stop automatic updates?
    - I tried setting `SERVERLESS_FRAMEWORK_FORCE_UPDATE=false` in the environment variable, but the automatic update did not stop. If you want to fix the version used by `serverless deploy`, is it only possible to specify `frameworkVersion` in `serverless.yml`?
2. Please tell me all the timing and conditions under which automatic updates run.
    - I think that automatic updates occur when you run a `serverless update` and every 24 hours (when you run any serverless commands after the updateLastChecked time recorded in `~/.serverless/binaries/metadata.json` has passed), is that correct? And, are there any other times when automatic updates occur?
3. What specifically happens if `SERVERLESS_FRAMEWORK_FORCE_UPDATE=true` is set?
4. What specifically happens if `SERVERLESS_FRAMEWORK_FORCE_UPDATE=false` is set?
5. Does the Serverless Framework's automatic update mechanism ignore the version specification in Node.js npm's package.json?

**Comment by ethanherbertson** [comment]
I would also like this clarified ASAP. It's kind of distressing that a tool that could be managed as a devDependency of a node project, using standard semver and npm/yarn/etc tooling, is now in the business of managing _itself_.

I was expecting the migration to v4 to include very few breaking changes, but this discovery will probably require a significant posture change for my team. Ever since an incompatibility in (as I recall) v1.18 led us to have to use different versions of Serverless in different projects (albeit temporarily), we've been happily just including Serverless as a dev dependency of each project we were using it in, and writing little one-liner scripts in package.json to run deploys and teardowns using the locally-installed version. That was such a pleasant (and self-documenting!) experience that we never went back to using global installs.

Now we're going to have to reconsider all of that in light of these auto-updates, and step one of that will require a MUCH clearer description of what is actually being updated and when. (And where!)

**Comment by Hiroki-Aoki** [comment]
> 6. When pinning the Serverless Framework version with [frameworkVersion in serverless.yml](https://www.serverless.com/framework/docs/providers/aws/guide/services#pinning-a-version), is there any way to reference the value in an external file? (I'd like to write the version in one configuration file and reference it from serverless.yml for multiple CloudFormation stacks.)

I create a new issue about regarding question 6 above. https://github.com/serverless/serverless/issues/12984

**Comment by Hiroki-Aoki** [comment]
Could you please answer my questions? I really hope your response.

(We understand that pinning versions is not recommended by developers of Serverless Framework and that automatically updating versions can bring benefits such as automating security responses, but due to constraints on our team's development process, it is difficult to allow automatic version updates. So we'd like to know the detailed specifications of automatic updates and pinning the version.)

**Comment by czubocha** [comment]
> Does it only update when you run some serverless command such as serverless deploy or serverless print, except when you run the serverless update command? (I don't think it updates periodically in the background like cron, is that right?)

Yes, that’s correct. It updates only when you run a Serverless command. It does not update in the background.

> Is this for when you have already installed Serverless Framework with npm install? What I mean is... I think the version of the binary installer (binary.js, etc.) changes depending on the version of Serverless Framework in package.json, so I think it's safe to make the version of Serverless Framework in package.json and the version of frameworkVersion in serverless.yml consistent. Is it correct?
> 
> When the version of Serverless Framework in package.json changes, I think the files under node_modules/serverless/ after npm install will change. And I think the files downloaded after npm install are the contents of the [repository on npm](https://www.npmjs.com/package/serverless?activeTab=code), not the [repository on GitHub](https://github.com/serverless/serverless). If so, for example, in the future, when the following parts of binary.js change in Serverless Framework 4.5.0, 4.6.0, etc., won't that affect the files that are finally downloaded under ~/.serverless/releases/x.x.x? I'm wondering if '0.0.2' may become '0.0.3' or higher.

You’re correct. While we strive to maintain compatibility between older Serverless Installer versions and newer Serverless Framework versions, the best approach is to keep the versions in `package.json` and `frameworkVersion` consistent. This ensures compatibility and avoids potential issues.

**Comment by Hiroki-Aoki** [comment]
@czubocha Thank you for your detailed explanation. I understand. Let me ask one more question.

There is a `node_modules` directory and `package.json` under `~/.serverless/releases/x.x.x/`. Does this mean that the Serverless Framework binary (in my case: `serverless-linux-amd64-0.0.2`) is running `npm install` on my local PC? Or is `npm install` not running on my local PC, and `node_modules` with contents already included are installed?

**Comment by henrist** [comment]
@czubocha I checked how all the recent releases would behave when it comes to pinning. Please describe why we see the following behavior, as this seems very wrong:

- `frameworkVersion: "4.4.x"` <-- gives 4.4.19
- `frameworkVersion: "4.4.19"` <-- gives 4.4.19
- `frameworkVersion: "4.4.20"` <-- gives 4.6.1 ❗❗ (it seems this has now become effectively `latest`)
- `frameworkVersion: "4.5.0"` <-- gives 4.5.4 ❗ (but this got latest patch on same minor)
- `frameworkVersion: "4.5.1"` <-- gives 4.5.1
- `frameworkVersion: "4.5.2"` <-- gives 4.5.2
- `frameworkVersion: "4.5.3"` <-- gives 4.5.3

This makes pinning completely unreliable.

**Comment by austencollins** [comment]
Thanks @henrist we are working on this right now.

**Comment by patrickbdaly** [comment]
How many versions will be automatically installed? Will it auto-remove older versions? Does it look at disk space? I am concerned over time this will grow significantly:

```
 du -sh -- *

212M	4.4.10
212M	4.4.11
 80M	4.4.12
194M	4.4.13
150M	4.4.14
 80M	4.4.15
 80M	4.4.16
 80M	4.4.18
 80M	4.4.19
 80M	4.4.6
142M	4.4.7
100M	4.5.1
103M	4.5.2
103M	4.6.0
103M	4.6.1
103M	4.6.2
```

**Comment by nicwise** [comment]
Just come across this @austencollins as my CICD pipeline failed part way thru when 4.14.2 was released. We baked in 4.14.1

I find this TOTALLY unacceptable. Fine putting a warning in ("you're out of date!") and a command to update, but I have pipelines which are baked to a SPECIFIC version. I don't want it pulling from NPM (and failing in this case with the super helpful `Error: exit status 1`, likely an NPM issue) in CICD.

We have about 50 serverless.yml files over 10 repos, so I dont really want to maintain `frameworkVersion` across all of them. Thats just crazy.

Can I suggest Serverless engineering have a close read of https://12factor.net/

I'm right on the cusp of "serverless is useful enough that it's a lot of work to roll my own via terraform" but this might push us over. Lack of one build + mix in config for prod/dev was close...

**Comment by nicwise** [comment]
... and support says "we dont release often"

3 releases in 3 weeks. every one breaks our build until we rebuild the container (deploying / running database migrations this time! What fun!)
